### ISSUE TITLE
Use of `size_t` for File Length Limits Usefulness of LFS in 32-bit Processes

### ISSUE THREAD

**Issue by whoisj** [libgit2/libgit2#3112]
The `size_t` type is defined (at least on Windows) as 

```
#ifdef _WIN64
    typedef unsigned __int64 size_t;
#else
    typedef unsigned int     size_t;
#endif
```

This limits any value of `size_t` to 4GB, which we know is not true when dealing the kinds of files LFS is designed to manage. There are a number of places throughout the libgit2 code base where a type like `git_off_t` could be used to enforce 64-bit length support. However, in structs like `git_rawobj` this introduces problems, mostly due to the code churn required to support a 64-bit `len` field.

```
/* DO NOT EXPORT */
typedef struct {
	void *data;			/**< Raw, decompressed object data. */
	size_t len;			/**< Total number of bytes in data. */
	git_otype type;		/**< Type of this object. */
} git_rawobj;
```

Basically there are three options that I see:
 1. Fix up all file related code to use `git_off_t` and cast to `size_t` when interfacing with extern API which accept `size_t`.
 2. Same as above but add a new unsigned 64 bit type like `git_size_t`.
 3. Do nothing and let 32 bit builds of libgit2 fail with large files.

**Comment by ethomson** [comment]
It's not clear to me that we need to replace `size_t` throughout.  For example, `git_rawobj` is an internal type and handling a 64-bit length on a 32 bit machine is probably not something that we should care about, since we're about to load the entirety of that file into the `data` member, which is going to cause other problems on a 32 bit machine.  Is that roughly accurate?

It sounds like the big problem here is `git_index_entry` being a 32 bit unsigned value on 32 bit machines, is that correct?  Can we update only `git_index_entry`?

That said, this may be a nice opportunity to fix some of the problems with `git_off_t` and `size_t` that have accumulated where we assume that we can always cast them back and forth.

**Comment by carlosmn** [comment]
A `size_t` is correctly limited to 4GB in a 32-bit system, as it describes the possible size of a buffer, and you run out of address space at that point. This is why object sizes are this type, as we always hold them in memory; you simply cannot have a `git_rawobj` which is larger than `size_t`  (whether that's a good thing is a related by somewhat different discussion).

With the current architecture we cannot push in files larger than `SIZE_MAX`, which is definitely a bug, tracked in #2936 and I have been thinking about how to support reading files larger than `SIZE_MAX` but the Git formats are frankly not very supportive of this goal. I brought this up at Git Merge, tl;dr is that this is hard, because Git assumes you can fit stuff into memory so if you get a deltified object, you're basically out of luck. We could stream full objects and loose objects easily enough, but anything deltified is at best an absolute pain in the arse to handle, though maybe we can attempt to implement on-the-fly delta application for files we can't fit in memory.

There are a few places where a `git_off_t` would be a more correct choice than `size_t` but it's an unfortunate fact that we have to fit a lot of this stuff in memory so we just suck on 32 bits (though there are steps to be suck somewhat less).

**Comment by carlosmn** [comment]
@ethomson the index' entries are there for detecting changes rather than to be an authoritative source of information. The time and size are trimmed to 32 bits because you can detect a change by just keeping be bottom bytes.

We could change the size in the index entry to be `git_off_t` but it's still going to be wrong (on any machine) when you call `git_index_read()` so you shouldn't be using the index as the authoritative source for an object's size.

**Comment by ethomson** [comment]
> The time and size are trimmed to 32 bits because you can detect a change by just keeping be bottom bytes.

Yes, you're absolutely correct:

https://github.com/libgit2/libgit2/blob/master/src/index.c#L70

**Comment by whoisj** [comment]
@carlosmn @ethomson 
Seems like this is a non-issue then. I didn't realize it was incorrect to look at the index as authoritative. My mistake.

**Comment by carlosmn** [comment]
This aspect is something which we should probably document much more clearly as it can lead to unexpected results if a programmer isn't mindful of this.

When I formulated this a few days ago, one of the first things that came to mind was that I've probably made this mistake in my own code as well.
